### DISCUSSION TITLE
nested create

### DISCUSSION THREAD

**Discussion by hrithiqball** [prisma/prisma#22847] [Q&A]
### Question

I know in the docs said that nested query are not allowed in prisma and i just want to know how are you guys dealing with this. Here is the [doc link](https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries#nested-writes) in case anyone is wondering which part. This is just a discussion and sharing session

### How to reproduce (optional)

```ts
export async function createMaintenanceLibrary(
  maintenanceLibrary: CreateMaintenanceLibrary,
) {
  try {
    const newMaintenanceLibrary = await db.maintenanceLibrary.create({
      data: {
        title: maintenanceLibrary.title,
        description: maintenanceLibrary.description,
        createdById: maintenanceLibrary.createdById,
        updatedById: maintenanceLibrary.updatedById,
      },
    });

    maintenanceLibrary.checklistLibrary.forEach(async checklist => {
      const newChecklistLib = await db.checklistLibrary.create({
        data: {
          title: checklist.title,
          description: checklist.description,
          createdById: checklist.createdById,
          updatedById: checklist.updatedById,
          maintenanceLibraryId: newMaintenanceLibrary.id,
        },
      });

      checklist.taskLibrary.forEach(async task => {
        await db.taskLibrary.create({
          data: {
            taskActivity: task.taskActivity,
            description: task.description,
            listChoice: task.listChoice,
            taskType: task.taskType,
            createdById: task.createdById,
            updatedById: task.updatedById,
            checklistLibraryId: newChecklistLib.id,
            subtaskLibrary: {
              createMany: {
                data: task.subtaskLibrary,
              },
            },
          },
        });
      });
    });

    console.log(newMaintenanceLibrary);
    return newMaintenanceLibrary;
  } catch (error) {
    console.error(error);
    throw error;
  }
}
```

### Information about Prisma Schema, Client Queries and Environment (optional)
```prisma
model MaintenanceLibrary {
  id               String             @id @default(cuid())
  title            String
  description      String?
  createdById      String             @map("created_by_id")
  createdOn        DateTime           @default(now()) @map("created_on")
  updatedById      String             @map("updated_by_id")
  updatedOn        DateTime           @default(now()) @map("updated_on")
  checklistLibrary ChecklistLibrary[]
  createdBy        User               @relation("maintenance_library_created_by_idTouser", fields: [createdById], references: [id], onDelete: Cascade)
  updatedBy        User               @relation("maintenance_library_updated_by_idTouser", fields: [updatedById], references: [id], onDelete: Cascade)

  @@map("maintenance_library")
  @@schema("public")
}
```

**Comment by ludralph** [comment]
Hi @hrithiqball 👋 

Thank you for raising this question.

Nested queries are allowed in Prisma, and they are often used to create or update records along with their related records in a single transaction. This feature provides transactional guarantees for creating, updating, or deleting data across multiple tables within a single Prisma Client query. If any part of the query fails, Prisma Client rolls back all changes to maintain data integrity. The part of the [documentation](https://www.prisma.io/docs/orm/prisma-client/queries/relation-queries#nested-writes) you shared also confirms this.

**Reply by hrithiqball** [reply]
Yes I understand it could create a *single* nested. However, when I tried to nest it like my code purpose for example 1 maintenance create multiple checklist and *each* checklist create multiple task. and *each* task create more. It will definitely fail to do it.

**Reply by hrithiqball** [reply]
![image](https://github.com/prisma/prisma/assets/87375380/234c3848-5199-4ebc-bf4c-22b37bec5701)

This is the part of documentation that supports my statement and yes I did create a nested query in my code but only on last level. 
```ts
subtaskLibrary: {
              createMany: {
                data: task.subtaskLibrary,
              },
            }
```

**Reply by ludralph** [reply]
Hi @hrithiqball 

Thanks for confirming. Did you see the workaround that was mentioned in that section of the documentation as well.

> As a workaround, you can send a query to create the records that will be connected first, and then create the actual records. For example:
```js
const categories = await prisma.category.createManyAndReturn({
  data: [
    { name: 'Fun', },
    { name: 'Technology', },
    { name: 'Sports', }
  ],
  select: {
    id: true
  }
});

const posts = await prisma.post.createManyAndReturn({
  data: [{
    title: "Funniest moments in 2024",
    categoryId: categories[0].id
  }, {
    title: "Linux or macOS — what's better?",
    categoryId: categories[1].id
  },
  {
    title: "Who will win the next soccer championship?",
    categoryId: categories[2].id
  }]
});
```

**Reply by hrithiqball** [reply]
its been 1 year, i dont even remember the problem anymore. just hacky way to prod lol

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
